sdk/src/azure/core/az_mqtt5_rpc_client_codec.c

```diff
@@ -1,83 +1,420 @@
+        i += az_span_size(az_mqtt5_rpc_service_id_key) - 1;
+      }
+      else if (i == executor_id_index)
+      {
+        remainder = az_span_copy_u8(remainder, '+');
+        i += az_span_size(az_mqtt5_rpc_executor_id_key) - 1;
+      }
+      else if (i == command_id_index)
+      {
+        remainder = az_span_copy_u8(remainder, '+');
+        i += az_span_size(az_mqtt5_rpc_command_id_key) - 1;
+      }
+    }
+    else
+    {
+      remainder = az_span_copy_u8(remainder, c);
+    }
+  }
+
+  az_span_copy_u8(remainder, '\0');
+
+  if (out_mqtt_topic_length)
+  {
+    *out_mqtt_topic_length = (size_t)required_length;
+  }
+
+  return AZ_OK;
 }
 
-AZ_NODISCARD az_result az_mqtt5_rpc_client_codec_get_request_topic(
+AZ_NODISCARD az_result az_mqtt5_rpc_client_codec_parse_received_topic(
     az_mqtt5_rpc_client_codec* client,
-    az_span server_client_id,
-    az_span command_name,
-    az_span out_request_topic)
+    az_span received_topic,
+    az_mqtt5_rpc_client_codec_request_response* out_response)
 {
-  return az_rpc_get_topic_from_format(
-      client->_internal.options.request_topic_format,
-      client->_internal.model_id,
-      server_client_id,
-      client->_internal.client_id,
-      _az_span_is_valid(client->_internal.command_name, 1, 0) ? client->_internal.command_name
-                                                              : command_name,
-      out_request_topic,
-      NULL);
+  _az_PRECONDITION_NOT_NULL(client);
+  _az_PRECONDITION_VALID_SPAN(received_topic, 1, false);
+  _az_PRECONDITION_NOT_NULL(out_response);
+
+  az_span topic_format = client->_internal.options.subscription_topic_format;
+  int32_t invoker_client_index = az_span_find(topic_format, az_mqtt5_rpc_client_id_key);
+  int32_t model_id_index = az_span_find(topic_format, az_mqtt5_rpc_service_id_key);
+  int32_t executor_id_index = az_span_find(topic_format, az_mqtt5_rpc_executor_id_key);
+  int32_t command_id_index = az_span_find(topic_format, az_mqtt5_rpc_command_id_key);
+
+  int32_t format_idx = 0;
+  uint8_t format_char;
+  for (int32_t i = 0; i < az_span_size(received_topic); i++)
+  {
+    uint8_t c = az_span_ptr(received_topic)[i];
+    if (format_idx < az_span_size(topic_format))
+    {
+      format_char = az_span_ptr(topic_format)[format_idx];
+      if (format_char == '{')
+      {
+        if (format_idx == invoker_client_index
+            && (i + az_span_size(client->_internal.client_id) - 1) < az_span_size(received_topic))
+        {
+          az_span extracted_client_id
+              = az_span_slice(received_topic, i, i + az_span_size(client->_internal.client_id));
+          if (az_span_is_content_equal(extracted_client_id, client->_internal.client_id))
+          {
+            i += az_span_size(client->_internal.client_id) - 1;
+            format_idx += az_span_size(az_mqtt5_rpc_client_id_key) - 1;
+          }
+          else
+          {
+            return AZ_ERROR_IOT_TOPIC_NO_MATCH;
```

Comment:
Should this be an IoT specific error? Would this work right if the IoT portion of the sdk wasn't compiled?

Reply:
I believe we'll be using this error for now, I asked @CIPop about creating a `AZ_ERROR_MQTT_TOPIC_NO_MATCH` but he suggested to keep using the IoT specific error

Reply:
If we chose to keep things in the MQTT namespace it's likely we'll need another constant but I think most of these functions should be in a separate namespace: either IoT or something else.